[SPARK-12708][UI] Sorting task error in Stages Page when yarn mode.

yoshidakuy opened the pull request:
If sort column contains slash(e.g. "Executor ID / Host") when yarn mode,sort fail with following message.

![spark-12708](https://cloud.githubusercontent.com/assets/6679275/12193320/80814f8c-b62a-11e5-9914-7bf3907029df.png)

Ｉt's similar to SPARK-4313 .

Commit:
[SPARK-12708][UI] Sorting task error in Stages Page when yarn mode

AmplabJenkins commented:
Can one of the admins verify this patch?

sarutak commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -100,7 +101,19 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
       val taskPage = Option(parameterTaskPage).map(_.toInt).getOrElse(1)
-      val taskSortColumn = Option(parameterTaskSortColumn).getOrElse("Index")
+      val taskSortColumn = Option(parameterTaskSortColumn).map {
+        sortColumn =>
```
`sortColumn` should be on previous line like `.map { sortColumn =>`

sarutak commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -100,7 +101,19 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+        sortColumn =>
+          // If sortColumn contains "/", `getParameter("task.sort")` will return 
+          // "%252F" when yarn mode. we need additional decode.
+          // See also SPARK-4313, YARN-2844.
```
How about say like as follows?

```
// SPARK-12708
// Due to YARN-2844, "/" in the url will be encoded to "%252F" when
// running in yarn mode. `request.getParameter("task.sort")` will return
// "%252F". Therefore we need to decode it until we get the real id.
// SPARK-4313 is similar to this issue.
```

sarutak commented:
ok to test.

sarutak commented:
@yoshidakuy Does this issue affect both `yarn-client` and `yarn-cluster` right?
Also, could you show the screen shot after applying this patch.

SparkQA commented:
**[Test build #49013 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49013/consoleFull)** for PR 10663 at commit [`fc45425`](https://github.com/apache/spark/commit/fc45425bc80a039994e57d2d98c0a6f281427654).

SparkQA commented:
**[Test build #49013 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49013/consoleFull)** for PR 10663 at commit [`fc45425`](https://github.com/apache/spark/commit/fc45425bc80a039994e57d2d98c0a6f281427654).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49013/
Test FAILed.

Commit:
fixed PR comments

SparkQA commented:
**[Test build #49018 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49018/consoleFull)** for PR 10663 at commit [`62c04ca`](https://github.com/apache/spark/commit/62c04ca239d160d24bfb55dee6d2e827b8183f9a).

yoshidakuy commented:
@sarutak Thanks. I have addressed your comment.

> Does this issue affect both yarn-client and yarn-cluster right?

Yes.
Fixed screenshot is following.
![spark-12708-2](https://cloud.githubusercontent.com/assets/6679275/12198261/a583aa56-b651-11e5-8cd8-1289b7b5a408.PNG)

SparkQA commented:
**[Test build #49018 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49018/consoleFull)** for PR 10663 at commit [`62c04ca`](https://github.com/apache/spark/commit/62c04ca239d160d24bfb55dee6d2e827b8183f9a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49018/
Test PASSed.

zsxwing commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -100,7 +101,20 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+          // running in yarn mode. `request.getParameter("task.sort")` will return
+          // "%252F". Therefore we need to decode it until we get the real column name.
+          // SPARK-4313 is similar to this issue.
+          var column = sortColumn
```
Since this is very similar to `ExecutorThreadDumpPage`, could you create a new method in UIUtils (maybe call it decodeURLParameter and add comments to explain it) and update here and ExecutorThreadDumpPage to use it?

JoshRosen replied:
+1; good to reduce duplication here.

sarutak replied:
Make sense. 
@yoshidakuy can you add a new method to `UIUtils` to reduce the similar logic.
Then, call it from `ExecutorThreadDumpPage` and `StagePage`.
Also, could you add a test case for the new method to `UIUtilsSuite`?

yoshidakuy commented:
Thanks for comments and I agree. will fix later.

Commit:
Merge remote branch 'upstream/master' into SPARK-12708

Commit:
Fix PR comments and similar issue in PoolPage.

JoshRosen commented:
Ping @yoshidakuy, could you update this pull request to address our comments? It would be great to merge this soon.

JoshRosen commented:
(I think we're going to do a 1.6.1 soon, so if we get this in soon it can be in that release)

yoshidakuy commented:
@JoshRosen @sarutak Sorry for late.
I found same issue in PoolTable, so I fixed it too.

SparkQA commented:
**[Test build #49396 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49396/consoleFull)** for PR 10663 at commit [`3d508c6`](https://github.com/apache/spark/commit/3d508c6f6c69c2b2e4b29f8b9edf7048694e7f23).

SparkQA commented:
**[Test build #49396 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49396/consoleFull)** for PR 10663 at commit [`3d508c6`](https://github.com/apache/spark/commit/3d508c6f6c69c2b2e4b29f8b9edf7048694e7f23).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49396/
Test PASSed.

sarutak commented on core/src/main/scala/org/apache/spark/ui/exec/ExecutorThreadDumpPage.scala:
```diff
@@ -30,19 +29,7 @@ private[ui] class ExecutorThreadDumpPage(parent: ExecutorsTab) extends WebUIPage
-        }
-        id
-    }.getOrElse {
+    val executorId = Option(request.getParameter("executorId")).getOrElse {
```
`decodeURLParameter` is not invoked.

We should say like as follows right?

```
Option(...).map { executorId =>
  UIUtils.decodeURLParameter(executorId)
}.getOrElse {
  ...
}
```

sarutak commented on core/src/test/scala/org/apache/spark/ui/UIUtilsSuite.scala:
```diff
@@ -67,6 +67,20 @@ class UIUtilsSuite extends SparkFunSuite {
+    val decoded2 = "<driver>"
+
+    assert(decoded1.sameElements(decodeURLParameter(encoded1)))
+    assert(decoded2.sameElements(decodeURLParameter(encoded2)))
```
Why don't you just compare by `===` instead of `sameElements`?

zsxwing commented on core/src/main/scala/org/apache/spark/ui/jobs/PoolPage.scala:
```diff
@@ -31,7 +31,7 @@ private[ui] class PoolPage(parent: StagesTab) extends WebUIPage("pool") {
   def render(request: HttpServletRequest): Seq[Node] = {
     listener.synchronized {
-      val poolName = request.getParameter("poolname")
+      val poolName = UIUtils.decodeURLParameter(request.getParameter("poolname"))
```
NullPointerException if `poolname` is not set. Could you use `Option....getOrElse` like other place?

zsxwing commented on core/src/main/scala/org/apache/spark/ui/jobs/PoolTable.scala:
```diff
@@ -59,7 +61,8 @@ private[ui] class PoolTable(pools: Seq[Schedulable], parent: StagesTab) {
     val href = "%s/stages/pool?poolname=%s"
-      .format(UIUtils.prependBaseUri(parent.basePath), p.name)
+      .format(UIUtils.prependBaseUri(parent.basePath),
+        URLEncoder.encode(p.name, "UTF-8"))
```
This two lines can be fit into one line.

koyoshida committed:
Fixed PR comments.

SparkQA commented:
**[Test build #49407 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49407/consoleFull)** for PR 10663 at commit [`384b990`](https://github.com/apache/spark/commit/384b990e237d6c0b311042301ed0252dc689569f).

yoshidakuy commented:
@sarutak @zsxwing Thanks for your review. I have addressed your comments. Could you please check it?

zsxwing commented:
LGTM

SparkQA commented:
**[Test build #49407 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49407/consoleFull)** for PR 10663 at commit [`384b990`](https://github.com/apache/spark/commit/384b990e237d6c0b311042301ed0252dc689569f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49407/
Test FAILed.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #49414 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49414/consoleFull)** for PR 10663 at commit [`384b990`](https://github.com/apache/spark/commit/384b990e237d6c0b311042301ed0252dc689569f).

SparkQA commented:
**[Test build #49414 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49414/consoleFull)** for PR 10663 at commit [`384b990`](https://github.com/apache/spark/commit/384b990e237d6c0b311042301ed0252dc689569f).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49414/
Test FAILed.

sarutak commented:
retest this please.

SparkQA commented:
**[Test build #49433 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49433/consoleFull)** for PR 10663 at commit [`384b990`](https://github.com/apache/spark/commit/384b990e237d6c0b311042301ed0252dc689569f).

SparkQA commented:
**[Test build #49433 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49433/consoleFull)** for PR 10663 at commit [`384b990`](https://github.com/apache/spark/commit/384b990e237d6c0b311042301ed0252dc689569f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49433/
Test PASSed.

sarutak commented:
LGTM as well. Merging into `master` and `branch-1.6`. Thanks @yoshidakuy !
